[SMALLFIX] Clean up block writer when failed to cache data read from UFS

apc999 opened the pull request.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/17021/
Test PASSed.

aaudiber commented on core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java:
```diff
@@ -233,9 +232,8 @@ public int transferTo(ByteBuf buf) throws IOException {
-        LOG.warn("Failed to cache data read from UFS: %s", e.getMessage());
-        mBlockWriter.close();
-        mBlockWriter = null;
+        LOG.warn("Failed to cache data read from UFS in Alluxio: {}", e.getMessage());
```
The two warnings are the same, so it's hard to know which one caused a given log message. Could we make them specific enough to be unique?

calvinjia replied:
I think we need to request space here too.

apc999 replied:
done

aaudiber commented on core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java:
```diff
@@ -305,18 +303,16 @@ private void updateUnderFileSystemInputStream(long offset) throws IOException {
-      }
+      mBlockWriter.close();
+      mBlockWriter = null;
+      mLocalBlockStore.abortBlock(mBlockMeta.getSessionId(), mBlockMeta.getBlockId());
```
is the exception handling below still appropriate for the new call sites?

apc999 replied:
I think so.

aaudiber commented on core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java:
```diff
@@ -230,12 +229,13 @@ public int transferTo(ByteBuf buf) throws IOException {
-        LOG.warn("Failed to cache data read from UFS: %s", e.getMessage());
-        mBlockWriter.close();
-        mBlockWriter = null;
+        LOG.warn("Failed to cache data read from UFS on transferTo(): {}", e.getMessage());
```
what does transferTo() mean in terms that a user unfamiliar with the code would understand?

apc999 replied:
well essentially "read" is the same as "transferTo" but with different argument type.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/17023/
Test PASSed.

apc999 commented:
@yupeng9 there actually is an integration test covering this case but not really working (failed to catch the original bug) due to some mockito issue. I fixed the test (and added one). Tested that without my change, the tests will fail.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/17030/
Test PASSed.
